### PR TITLE
Shorten critical path between upload url creation and utilization.

### PR DESCRIPTION
**Pull request by arangatang**
## Problem
There is a chance for large uploads to Q Code Transform that the UploadUrl expires before its used. 
One of the reasons, particularly for large artifacts on HDDs, can be the reading from disk and calculating hash on the loaded zip. 

## Solution

This PR shortens the critical path by:
1. Only read file from disk once (was three times)
2. Only calculate hash once (was twice)
3. Loads file and calculates hash before the upload url is requested instead of after.

Testing: started a new transformation job, it passed the initial build.

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by damntrecky — APPROVED** [review]

File: packages/core/src/codewhisperer/commands/startTransformByQ.ts [hunk 1]
```diff
@@ -284,7 +284,8 @@ export async function preTransformationUploadCode(userInputState: UserInputState
         await vscode.commands.executeCommand('aws.amazonq.refresh') // so that button updates
         uploadId = await uploadPayload(payloadFilePath)
     } catch (error) {
-        const errorMessage = 'Failed to upload code'
+        getLogger().error(`${error}`)
+        const errorMessage = `Failed to upload code due to ${error}`
```
Comment (mk-fan) [inline packages/core/src/codewhisperer/commands/startTransformByQ.ts:288]: Should this be `error.message`?

File: packages/core/src/codewhisperer/service/transformByQHandler.ts [hunk 2]
```diff
@@ -237,17 +237,19 @@ export function getHeadersObj(sha256: string, kmsKeyArn: string | undefined) {
     return headersObj
 }
 
-// Consider enhancing the S3 client to include this functionality
-export async function uploadArtifactToS3(fileName: string, resp: CreateUploadUrlResponse) {
-    const sha256 = getSha256(fileName)
-    const headersObj = getHeadersObj(sha256, resp.kmsKeyArn)
-
+// TODO: later, consider enhancing the S3 client to include this functionality
```
Comment (mk-fan) [inline packages/core/src/codewhisperer/service/transformByQHandler.ts:240]: nit: As discussed before, we should avoid adding more TODOs.

**Review by dhasani23 — APPROVED** [review]

File: packages/core/src/codewhisperer/service/transformByQHandler.ts (hunk 2 above)
Comment (dhasani23) [reply packages/core/src/codewhisperer/service/transformByQHandler.ts:240]: This is fine for now, I'm planning on going through and making SIMs for all TODOs

File: packages/core/src/codewhisperer/commands/startTransformByQ.ts (hunk 1 above)
Comment (dhasani23) [reply packages/core/src/codewhisperer/commands/startTransformByQ.ts:288]: I think either is fine, probably don't need the back ticks and $, { though

Comment (damntrecky) [reply packages/core/src/codewhisperer/commands/startTransformByQ.ts:288]: Question/Suggestion: `getLogger().error(error)` or `getLogger().error(error.message)` should work

Comment (arangatang) [reply packages/core/src/codewhisperer/commands/startTransformByQ.ts:288]: Moved it one line down to reuse the `errorMessage` variable instead, keeps it consistent.

File: packages/core/src/codewhisperer/service/transformByQHandler.ts (hunk 2 above)
Comment (arangatang) [reply packages/core/src/codewhisperer/service/transformByQHandler.ts:240]: Removed it.

**Comment by arangatang** [comment]
The failures does not seem to be related to my changes. There seem to be failures on `getWorkspaceFoldersByPrefixes` which I did not touch.

File: .changes/next-release/Bug Fix-3b5c1306-6cbd-4592-a73e-bc92120ddcf3.json [hunk 3]
```diff
@@ -0,0 +1,4 @@
+{
+	"type": "Bug Fix",
+	"description": "Shorten critical path between presigned url creation and usage to lower chance of url expiry"
```
Comment (justinmk3) [inline .changes/next-release/Bug Fix-3b5c1306-6cbd-4592-a73e-bc92120ddcf3.json:3]: Please review the changelog guidlines: https://github.com/aws/aws-toolkit-vscode/blob/master/CONTRIBUTING.md#changelog

File: packages/core/src/codewhisperer/commands/startTransformByQ.ts (hunk 1 above)
Comment (justinmk3) [reply packages/core/src/codewhisperer/commands/startTransformByQ.ts:287]: It should be `error.message`, as mentioned in the code guidelines: https://github.com/aws/aws-toolkit-vscode/blob/master/docs/CODE_GUIDELINES.md#exceptions

The stacktrace is minified, thus it's just noise (in both the logs and telemetry).

File: packages/core/src/codewhisperer/service/transformByQHandler.ts [hunk 4]
```diff
@@ -213,9 +213,9 @@ export async function validateOpenProjects(projects: vscode.QuickPickItem[]) {
     return projectsValidToTransform
 }
 
-export function getSha256(fileName: string) {
+export function getSha256(buffer: Buffer) {
     const hasher = crypto.createHash('sha256')
-    hasher.update(fs.readFileSync(fileName))
+    hasher.update(buffer)
     return hasher.digest('base64')
```
Comment (justinmk3) [inline packages/core/src/codewhisperer/service/transformByQHandler.ts:219]: does it matter if it's base64 or hex? If hex is acceptable you could use the existing util function: https://github.com/aws/aws-toolkit-vscode/blob/721e951fd291aebb47c89754b7314dd36ed01b5b/packages/core/src/shared/utilities/textUtilities.ts#L104

File: .changes/next-release/Bug Fix-3b5c1306-6cbd-4592-a73e-bc92120ddcf3.json
[the remark this answers is not in the document]
Comment (justinmk3) [reply .changes/next-release/Bug Fix-3b5c1306-6cbd-4592-a73e-bc92120ddcf3.json:1]: The PR problem describes something that sounds customer-impacting:

> There is a chance for large uploads to Q Code Transform that the UploadUrl expires before its used.
> One of the reasons, particularly for large artifacts on HDDs, can be the reading from disk and calculating hash on the loaded zip.

That's one reason why a "Problem" statement is required in PRs.

File: packages/core/src/codewhisperer/service/transformByQHandler.ts (hunk 4 above)
Comment (dhasani23) [reply packages/core/src/codewhisperer/service/transformByQHandler.ts:219]: It has to be `base64`. Could modify this `getStringHash` and update the ~6 uses in the codebase accordingly if we really wanted to.

File: packages/core/src/codewhisperer/commands/startTransformByQ.ts (hunk 1 above)
Comment (arangatang) [reply packages/core/src/codewhisperer/commands/startTransformByQ.ts:287]: Updated it.

File: .changes/next-release/Bug Fix-3b5c1306-6cbd-4592-a73e-bc92120ddcf3.json (hunk 3 above)
Comment (arangatang) [reply .changes/next-release/Bug Fix-3b5c1306-6cbd-4592-a73e-bc92120ddcf3.json]: Done.

File: .changes/next-release/Bug Fix-3b5c1306-6cbd-4592-a73e-bc92120ddcf3.json
[the remark this answers is not in the document]
Comment (arangatang) [reply .changes/next-release/Bug Fix-3b5c1306-6cbd-4592-a73e-bc92120ddcf3.json:1]: Updated the changelog.

**Comment by damntrecky** [comment]
> The failures does not seem to be related to my changes. There seem to be failures on `getWorkspaceFoldersByPrefixes` which I did not touch.

I'm having the same test failure in another PR -> https://github.com/aws/aws-toolkit-vscode/pull/4451

**Comment by damntrecky** [comment]
I believe this PR -> https://github.com/aws/aws-toolkit-vscode/pull/4531/files fixes it. Try rebasing.

**Review by nkomonen-amazon — APPROVED** [review]

**Pull request merged**